Python/flowgraph.c

```diff
@@ -1443,6 +1443,84 @@ optimize_if_const_list_or_set(PyObject *const_cache, cfg_instr* inst, int n, PyO
     return SUCCESS;
 }
 
+/*
+  Walk basic block upwards starting from "start" to collect instruction pair
+  that loads consts skipping NOP's in between.
+*/
+static bool
+find_load_const_pair(basicblock *bb, int start, cfg_instr **first, cfg_instr **second)
+{
+    cfg_instr *second_load_const = NULL;
+    while (start >= 0) {
+        cfg_instr *inst = &bb->b_instr[start--];
+        if (inst->i_opcode == NOP) {
+            continue;
+        }
+        if (!loads_const(inst->i_opcode)) {
+            return false;
+        }
+        if (second_load_const == NULL) {
+            second_load_const = inst;
+            continue;
+        }
+        *first = inst;
+        *second = second_load_const;
+        return true;
+    }
+    return false;
+}
+
+/* Determine opcode & oparg for freshly folded constant. */
+static int
+newop_from_folded(PyObject *newconst, PyObject *consts,
+                    PyObject *const_cache, int *newopcode, int *newoparg)
+{
+    if (PyLong_CheckExact(newconst)) {
+        int overflow;
+        long val = PyLong_AsLongAndOverflow(newconst, &overflow);
+        if (!overflow && val >= 0 && val < 256) {
```

Comment:
```suggestion
        if (!overflow && val >= 0 && val < 256 && val < _PY_NSMALLPOSINTS) {
```
This check appears in codegen.c `codegen_addop_load_const` as well. Shall we have a macro in a header somewhere?

Reply:
There is related PR to this: https://github.com/python/cpython/pull/129553. I could make macro for this as part of that PR and then use it here. In that case leave condition as it is and rename PR to indicate that we need to make condition a macro to reuse it in CFG. WDYT?

Reply:
Yes, it could be in this PR.

Reply:
> Yes, it could be in this PR.

You mean https://github.com/python/cpython/pull/129553, right? 😅

Reply:
Maybe in a future PR we (you?) could remove this whole small int things from codegen.c. Just emit LOAD_CONST there and then here in flowgraph.c determine whether it should be optimised to LOAD_SMALL_INT. Then we won't have this logic in two places.

Reply:
I will refer to this comment in the issue to keep track of it.

Reply:
> > Yes, it could be in this PR.
> 
> You mean #129553, right? 😅

No, sorry I meant it could be rolled into the current PR.

Reply:
Alright, should we close https://github.com/python/cpython/pull/129553 then?

Reply:
Or just leave it because we're not going to have it used in 2 placed later on.

Reply:
You mean to reuse that PR for:

>Maybe in a future PR we (you?) could remove this whole small int things from codegen.c. Just emit LOAD_CONST there and then here in flowgraph.c determine whether it should be optimised to LOAD_SMALL_INT. Then we won't have this logic in two places.

?

Reply:
Either reuse or close it and open a new one. Doesn't matter.